src/main/java/me/qoomon/gitversioning/GitVersioning.java

```diff
@@ -109,19 +68,67 @@ public static GitVersionDetails determineVersion(
             dataMap.put("version", currentVersion);
             dataMap.put("version.release", currentVersion.replaceFirst("-SNAPSHOT$", ""));
             return transformProperties(
-                    currentProperties, finalVersionDescription.getPropertyDescriptions(), dataMap);
+                    currentProperties, versioningInfo.description.getPropertyDescriptions(), dataMap);
         };
 
         return new GitVersionDetails(
                 repoSituation.isClean(),
                 repoSituation.getHeadCommit(),
-                gitRefType,
-                gitRefName,
+                versioningInfo.gitRefType,
+                versioningInfo.gitRefName,
                 versionTransformer,
                 propertiesTransformer
         );
     }
 
+    private static VersioningInfo getVersioningInfo(GitRepoSituation repoSituation, VersionDescription commitVersionDescription, List<VersionDescription> branchVersionDescriptions, List<VersionDescription> tagVersionDescriptions, boolean preferTags) {
+
+        // tags take precedence over branches
+        VersioningInfo versioningInfo = null;
+        if (preferTags) {
+            versioningInfo = getTagVersioningInfo(repoSituation, tagVersionDescriptions);
+        }
+        if (versioningInfo == null) {
+            versioningInfo = getBranchVersioningInfo(repoSituation, branchVersionDescriptions);
+        }
+        if (!preferTags && versioningInfo == null) {
+            versioningInfo = getTagVersioningInfo(repoSituation, tagVersionDescriptions);
+        }
```

Comment:
I think it is easier to read if we do it like this 
```
VersioningInfo versioningInfo = getBranchVersioningInfo(repoSituation, branchVersionDescriptions);

 if (versioningInfo == null || preferTags) {
    versioningInfo = getTagVersioningInfo(repoSituation, tagVersionDescriptions);
 }     
```

Reply:
I agree, although it still needs to be `(versioningInfo == null && !preferTags)` - your version will work, but if versioningInfo is null and preferTags is true, it will unsuccessfully try to get the tag versioning info again (it did that above already, and nothing changed).

Reply:
you're right, my bad :-D